### PR TITLE
32380 deprecate squeeze in groupby

### LINKED ISSUE #32380: Deprecate groupby() squeeze option
#### Code Sample

```python
import pandas as pd
print(pd.DataFrame([{
    'A': 1,
    'B': 1,
}, {
    'A': 2,
    'B': 2,
}, {
    'A': 2,
    'B': 3,
}]).groupby('A', squeeze=True).count())

```
#### Problem description

I expected `.groupby(squeeze=True)` to, well, squeeze, and `count()` to return a Series. Instead `squeeze=True` doesn't seem to do anything, and `count()` returns a DataFrame.

A workaround is to write `.groupby('A').count().squeeze()`, which does work.

#### Expected Output

```
A
1    1
2    2
Name: B, dtype: int64
```

#### Actual output

```
   B
A
1  1
2  2
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-4-amd64
machine          : x86_64
processor        :
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.0.1
numpy            : 1.17.4
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 18.1
setuptools       : 44.0.0
Cython           : None
pytest           : 4.6.9
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.0
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.1
IPython          : 7.12.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.5.0
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 4.6.9
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #32380
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I deprecated the squeeze keyword. I think I got every points this keyword was used. 
I hope that I considered every point necessary to deprecate a keyword. I tried to use other points were a keyword was deprecated as example.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -5813,26 +5814,35 @@ def groupby(
...
+                ),
+                FutureWarning,
+                stacklevel=2,
+            )
+
         if level is None and by is None:
             raise TypeError("You have to supply one of 'by' and 'level'")
         axis = self._get_axis_number(axis)
-
-        return DataFrameGroupBy(
```
Comment (mroeschke) [inline pandas/core/frame.py:5825]: Could you revert this assignment?

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -1253,34 +1253,7 @@ def first_not_none(values):
 
             if isinstance(v, (np.ndarray, Index, Series)):
                 if isinstance(v, Series):
-                    applied_index = self._selected_obj._get_axis(self.axis)
```
Comment (mroeschke) [inline pandas/core/groupby/generic.py:1256]: For deprecations, we can only remove this once we release pandas 2.0. `squeeze` must continue to work up until then, so we cannot remove all this code just yet.

**Review by mroeschke — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 2 above)
Comment (phofl) [reply pandas/core/groupby/generic.py:1256]: Oh sorry, I did not know that.

Just to get that right:
So I can only add a warning in a Pull Request, which will be merged into the master?

This Pull Request could be added before releasing 2.0 but not earlier?

Comment (mroeschke) [reply pandas/core/groupby/generic.py:1256]: > So I can only add a future warning in a Pull Request, which will be merged into the master?

Correct.

> This Pull Request could be merged into master before releasing 2.0 but not earlier?

This pull request with just the futurewarnings can be merged into master before releasing 2.0

So another pull request will need to be added in the future where this code is removed when we have decided that 2.0 will be the next release.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -5813,26 +5814,35 @@ def groupby(
...
+        grouped = DataFrameGroupBy(
             obj=self,
             keys=by,
             axis=axis,
             level=level,
             as_index=as_index,
             sort=sort,
             group_keys=group_keys,
-            squeeze=squeeze,
```
Comment (WillAyd) [inline pandas/core/frame.py:5833]: Want to revert this as well; just deprecate for now will remove later for pandas 2.0

File: pandas/core/generic.py [hunk 4]
```diff
@@ -7419,6 +7419,8 @@ def clip(
         squeeze : bool, default False
             Reduce the dimensionality of the return type if possible,
             otherwise return a consistent type.
+            deprecated:: 1.1.0
```
Comment (WillAyd) [inline pandas/core/generic.py:7422]:
```suggestion
            .. deprecated:: 1.1.0
```

Should put a newline before as well

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks for doing this - looks pretty good. As noted by @mroeschke looks like you did a little more work than was required; shouldn't remove anything just yet only warn the user appropriately

File: pandas/core/groupby/generic.py (hunk 2 above)
Comment (phofl) [reply pandas/core/groupby/generic.py:1256]: Thanks very much for the clarification. I will modify the PR accordingly tomorrow.

File: pandas/core/frame.py (hunk 3 above)
Comment (phofl) [reply pandas/core/frame.py:5833]: Yes, I will revert the commits and add only the warning for SeriesGroupBy and DataFrameGroupBy.

File: pandas/core/generic.py (hunk 4 above)
Comment (phofl) [reply pandas/core/generic.py]: Thanks, I did that in the newest commit

File: pandas/core/frame.py (hunk 3 above)
Comment (phofl) [reply pandas/core/frame.py]: I reverted the actual code changes. There is only a warning raised right now.

File: pandas/core/groupby/generic.py (hunk 2 above)
Comment (phofl) [reply pandas/core/groupby/generic.py]: I reverted the actual code changes and kept the warning.

File: pandas/core/frame.py (hunk 1 above)
Comment (phofl) [reply pandas/core/frame.py]: Done

File: pandas/core/generic.py [hunk 5]
```diff
@@ -7419,6 +7419,9 @@ def clip(
         squeeze : bool, default False
             Reduce the dimensionality of the return type if possible,
             otherwise return a consistent type.
+
+            deprecated:: 2.0
```
Comment (mroeschke) [inline pandas/core/generic.py:7423]: This will be deprecated in 1.1

File: pandas/core/frame.py [hunk 6]
```diff
@@ -5813,11 +5814,23 @@ def groupby(
...
-        squeeze: bool = False,
+        squeeze: bool = no_default,
         observed: bool = False,
     ) -> "DataFrameGroupBy":
         from pandas.core.groupby.generic import DataFrameGroupBy
 
+        if squeeze is not no_default:
+            warnings.warn(
+                (
+                    "The `squeeze` parameter in pd.groupby is deprecated and "
```
Comment (mroeschke) [inline pandas/core/frame.py:5825]: You can remove the `in pd.groupby`

File: pandas/core/series.py [hunk 7]
```diff
@@ -1646,11 +1647,23 @@ def groupby(
...
-        squeeze: bool = False,
+        squeeze: bool = no_default,
         observed: bool = False,
     ) -> "SeriesGroupBy":
         from pandas.core.groupby.generic import SeriesGroupBy
 
+        if squeeze is not no_default:
+            warnings.warn(
+                (
+                    "The `squeeze` parameter in pd.groupby is deprecated and "
```
Comment (mroeschke) [inline pandas/core/series.py:1658]: Same

**Comment by mroeschke** [comment]
We will still need a whatsnew entry to this (the one you had in a prior commit was good iirc)

File: pandas/core/frame.py (hunk 6 above)
Comment (phofl) [reply pandas/core/frame.py:5825]: Done

File: pandas/core/series.py (hunk 7 above)
Comment (phofl) [reply pandas/core/series.py]: Done

File: pandas/core/generic.py (hunk 5 above)
Comment (phofl) [reply pandas/core/generic.py]: Done

**Comment by phofl** [comment]
@mroeschke  I have added the whats new entry.

File: pandas/core/generic.py (hunk 4 above)
Comment (WillAyd) [reply pandas/core/generic.py:7423]: Did you push the latest changes? Doesn’t appear to be updated here on GitHub

Comment (phofl) [reply pandas/core/generic.py:7423]: If I view the changes over all commits, there is an empty line after the description and before deprecated: 1.1.0

Comment (mroeschke) [reply pandas/core/generic.py:7423]: Looks like this change still needs to be made

**Comment by mroeschke** [comment]
Looks good. Just need to merge master to fix the merge conflict and address @WillAyd's comment.

**Comment by phofl** [comment]
> 
> 
> Looks good. Just need to merge master to fix the merge conflict and address @WillAyd's comment.

Done

**Review by mroeschke — APPROVED** [review]

**Comment by topper-123** [comment]
Looks good. Can you update.

**Comment by phofl** [comment]
> 
> 
> Looks good. Can you update.

Done

**Review by topper-123 — APPROVED** [review]

**Comment by WillAyd** [comment]
Ci failure should be unrelated - can you merge master and repush?

**Comment by jreback** [comment]
thanks @phofl very nice.

**Pull request merged**